src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -510,22 +567,18 @@ public BoundLambda BindForReturnTypeInference(NamedTypeSymbol delegateType)
             return result;
         }
 
-        private static MethodSymbol GetCacheKey(NamedTypeSymbol delegateType)
+        private MethodSymbol GetCacheKey(NamedTypeSymbol delegateType)
         {
-            delegateType = delegateType.GetDelegateType();
-
-            if (delegateType != null)
+            var invoke = DelegateInvokeMethod(delegateType);
+            if ((object)invoke != null)
             {
-                var invoke = delegateType.DelegateInvokeMethod;
-                if (invoke != null)
-                {
-                    return invoke;
-                }
+                return invoke;
             }
 
             // delegateType or DelegateInvokeMethod can be null in cases of malformed delegates
-            // in such case we would want something trivial with no parameters, like a fake static ctor
-            return new SynthesizedStaticConstructor(delegateType);
+            // in such case we would want something trivial with no parameters, like a fake static ctor.
+            // Since the containingType of the .cctor must be non-null, System.Object is used.
+            return new SynthesizedStaticConstructor(binder.Compilation.GetSpecialType(SpecialType.System_Object));
```

Comment:
I can't find just now the place we're you're generating the cache. I had changed the cache to also key upon return type for ALL lambdas, even though keying upon return type is only needed for async lambdas. I worried that this would hurt perf in the common non-async case. Did you do the same? have the same worry?